File: Java/src/test/java/parrot/ParrotTest.java
package parrot;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class ParrotTest {

    @Test
    public void getSpeedOfEuropeanParrot() {
        Parrot parrot = newEuropeanParrot();
        assertEquals(12.0, parrot.getSpeed(), 0.0);
    }

    @Test
    public void getSpeedOfAfricanParrot_With_One_Coconut() {
        Parrot parrot = newAfricanParrot(1);
        assertEquals(3.0, parrot.getSpeed(), 0.0);
    }

    @Test
    public void getSpeedOfAfricanParrot_With_Two_Coconuts() {
        Parrot parrot = newAfricanParrot(2);
        assertEquals(0.0, parrot.getSpeed(), 0.0);
    }

    @Test
    public void getSpeedOfAfricanParrot_With_No_Coconuts() {
        Parrot parrot = newAfricanParrot(0);
        assertEquals(12.0, parrot.getSpeed(), 0.0);
    }

    @Test
    public void getSpeedNorwegianBlueParrot_nailed() {
        Parrot parrot = newFixedNorwegianBlueParrot();
        assertEquals(0.0, parrot.getSpeed(), 0.0);
    }

    @Test
    public void getSpeedNorwegianBlueParrot_not_nailed() {
        Parrot parrot = newMovableNorwegianBlueParrot(1.5);
        assertEquals(18.0, parrot.getSpeed(), 0.0);
    }

    @Test
    public void getSpeedNorwegianBlueParrot_not_nailed_high_voltage() {
        Parrot parrot = newMovableNorwegianBlueParrot(4);
        assertEquals(24.0, parrot.getSpeed(), 0.0);
    }

    private Parrot newEuropeanParrot() {
        return new EuropeanParrot();
    }

    private Parrot newAfricanParrot(int numberOfCoconuts) {
        return new AfricanParrot(numberOfCoconuts);
    }

    private Parrot newFixedNorwegianBlueParrot() {
        return new FixedNorwegianBlueParrot();
    }

    private Parrot newMovableNorwegianBlueParrot(double voltage) {
        return new MovableNorwegianBlueParrot(voltage);
    }
}

File: Kotlin/src/test/kotlin/parrot/ParrotTest.kt
package parrot

import org.junit.Test

import org.junit.Assert.assertEquals

class ParrotTest {

    @Test
    fun getSpeedOfEuropeanParrot() {
        val parrot = EuropeanParrot(12.0)
        assertEquals(12.0, parrot.speed(), 0.0)
    }

    @Test
    fun getSpeedOfAfricanParrot_With_One_Coconut() {
        val parrot = AfricanParrot( 1)
        assertEquals(3.0, parrot.speed(), 0.0)
    }

    @Test
    fun getSpeedOfAfricanParrot_With_Two_Coconuts() {
        val parrot = AfricanParrot(2)
        assertEquals(0.0, parrot.speed(), 0.0)
    }

    @Test
    fun getSpeedOfAfricanParrot_With_No_Coconuts() {
        val parrot = AfricanParrot(0)
        assertEquals(12.0, parrot.speed(), 0.0)
    }

    @Test
    fun getSpeedNorwegianBlueParrot_nailed() {
        val parrot = NorwegianBlueParrot(1.5, true)
        assertEquals(0.0, parrot.speed(), 0.0)
    }

    @Test
    fun getSpeedNorwegianBlueParrot_not_nailed() {
        val parrot = NorwegianBlueParrot(1.5, false)
        assertEquals(18.0, parrot.speed(), 0.0)
    }

    @Test
    fun getSpeedNorwegianBlueParrot_not_nailed_high_voltage() {
        val parrot = NorwegianBlueParrot(4.0, false)
        assertEquals(24.0, parrot.speed(), 0.0)
    }
}
